internal/bminventory/inventory.go

```diff
@@ -1239,30 +1309,45 @@ func (b *bareMetalInventory) GenerateInfraEnvISOInternal(ctx context.Context, in
 		return common.NewApiError(http.StatusInternalServerError, errors.New(msg))
 	}
 
-	if infraEnv, err = common.GetInfraEnvFromDB(b.db, infraEnv.ID); err != nil {
+	err = b.createAndUploadNewImage(ctx, log, infraEnv.ProxyHash, infraEnv.ID, infraEnv.Type, true, imageExists)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (b *bareMetalInventory) createAndUploadNewImage(ctx context.Context, log logrus.FieldLogger, infraEnvProxyHash string,
+	infraEnvID strfmt.UUID, imageType models.ImageType, v2 bool, imageExists bool) error {
+
+	infraEnv, err := common.GetInfraEnvFromDB(b.db, infraEnvID)
+	if err != nil {
 		log.WithError(err).Errorf("failed to get infra env %s after update", infraEnv.ID)
-		return common.NewApiError(http.StatusInternalServerError, err)
+		msg := "Failed to generate image: error fetching updated infra env metadata"
+		b.eventsHandler.AddEvent(ctx, infraEnvID, nil, models.EventSeverityError, msg, time.Now())
+		return err
 	}
 
-	if imageExists {
-		if err = b.updateImageInfoPostUpload(ctx, infraEnv, infraEnv.ProxyHash, infraEnv.Type, false, true); err != nil {
-			return common.NewApiError(http.StatusInternalServerError, err)
+	if imageExists && b.ImageServiceBaseURL == "" {
+		if err = b.updateImageInfoPostUpload(ctx, infraEnv, infraEnvProxyHash, imageType, false, v2); err != nil {
+			return err
 		}
 
-		log.Infof("Re-used existing InfraEnv <%s> image", infraEnv.ID)
+		log.Infof("Re-used existing image <%s>", infraEnv.ID)
+		b.eventsHandler.AddEvent(ctx, infraEnvID, nil, models.EventSeverityInfo,
+			fmt.Sprintf(`Re-used existing image rather than generating a new one (image type is "%s")`, imageType),
+			time.Now())
 		return nil
 	}
 
-	err = b.createAndUploadNewImage(ctx, log, infraEnv.ProxyHash, infraEnv, infraEnv.Type, true)
-	if err != nil {
-		return err
-	}
+	if b.ImageServiceBaseURL != "" {
+		if err = b.updateExternalImageInfo(infraEnv, infraEnvProxyHash, imageType); err != nil {
+			return err
+		}
 
-	return nil
-}
+		return nil
+	}
```

Comment:
Could you move this up above line number 1331, and then you wouldn't need `imageExists && b.ImageServiceBaseURL == ""`?

Reply:
:+1: